cpp/src/arrow/util/hashing.h

```diff
@@ -824,6 +898,13 @@ class BinaryMemoTable : public MemoTable {
     };
     return hash_table_.Lookup(h, cmp_func);
   }
+
+ public:
+  void MergeTable(BinaryMemoTable& other_table) {
+    other_table.VisitValues(0, [=](const util::string_view& other_value) {
+      ARROW_WARN_NOT_OK(this->MaybeInsert(other_value), "Merging BinaryMemoTable");
```

Comment:
We should probably return an error to the caller, not log a warning.

Reply:
made a JIRA to follow this up in the near-ish future: https://issues.apache.org/jira/browse/ARROW-17074